{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Seq Experimental Notebook"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "```bash\n",
    "# Location of tools' sources and executables\n",
    "export TOOLS=`pwd`/tools\n",
    "# Location of test datasets\n",
    "export DATA=`pwd`/data\n",
    "# Locatiion of the local Seq installation\n",
    "export SEQ=`pwd`/seq\n",
    "# Make sure to point LD_LIBRARY_PATH, PATH and SEQ_LIBRARY_PATH to your Seq installation directory\n",
    "export LD_LIBRARY_PATH=$SEQ/build:$SEQ/deps/lib:${LD_LIBRARY_PATH}\n",
    "export PATH=$SEQ/build:${PATH}\n",
    "export SEQ_LIBRARY_PATH=$SEQ/build\n",
    "# Use 1 thread\n",
    "export OMP_NUM_THREADS=1\n",
    "# Current working directory\n",
    "DIR=`pwd`\n",
    "# Make the necessary directories\n",
    "mkdir -p $DIR/runs/time\n",
    "mkdir -p $DIR/runs/stderr\n",
    "mkdir -p $DIR/runs/stdout\n",
    "mkdir -p $TOOLS/seq\n",
    "export OUT=$DIR/runs/out\n",
    "mkdir -p $OUT\n",
    "\n",
    "# Function that collects the performance metrics of each experiment.\n",
    "function tm {\n",
    "  TIME=`which time`  # use gtime on macOS\n",
    "  NAME=$1\n",
    "  if grep -qs 'Exit status: 0' $DIR/runs/time/$NAME ; then\n",
    "    echo -ne \"- $NAME already done \"\n",
    "    grep  'Exit status:' $DIR/runs/time/$NAME\n",
    "  else\n",
    "    cmd=\"$TIME -v -o $DIR/runs/time/$NAME ${@:2} >$DIR/runs/stdout/$NAME 2>$DIR/runs/stderr/$NAME\"\n",
    "    eval \"$cmd\"\n",
    "    es=$?\n",
    "    if [ $es -ne 0 ] ; then\n",
    "      echo \"- $NAME failed with $es: $cmd\"\n",
    "      cat $DIR/runs/stderr/$NAME\n",
    "    else\n",
    "      echo \"- $NAME completed\"\n",
    "    fi\n",
    "  fi\n",
    "}\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## CORA (Homology table construction)\n",
    "\n",
    "We have used [CORA 1.1.5b](https://github.com/denizy/cora/tree/6a53c7b9b25eb8c33d3de92f00481310c16fdd5c) to evaluate CORA performance.\n",
    "\n",
    "CORA first needs to be run with default parameters to generate `__temporary_CORA_files` directory.\n",
    "You can kill CORA once this directory is generated and proceed by running the commands below:\n",
    "\n",
    "```bash\n",
    "cd $OUT/cora\n",
    "# Generate inexact homology table\n",
    "tm cora-inexact $TOOLS/cora/homTable_setup $DATA/genome.fa 64 __temporary_CORA_files/__TEMP__Aux_Physical_Splits_File cora_hom_exact_k64 cora_hom_inexact_k64 __temporary_CORA_files/__TEMP__Aux_Parallel_Splits_File EXACT 1 FULL\n",
    "# Generate exact homology table (needs 2 steps)\n",
    "tm cora-exact-1 $TOOLS/cora/homTable_setup $DATA/genome.fa 64 __temporary_CORA_files/__TEMP__Aux_Physical_Splits_File cora_hom_exact_k64 cora_hom_inexact_k64 __temporary_CORA_files/__TEMP__Aux_Parallel_Splits_File EXACT 1 ONLY_EXACT_COMPACT\n",
    "tm cora-exact-2 $TOOLS/cora/homTable_setup $DATA/genome.fa 64 __temporary_CORA_files/__TEMP__Aux_Physical_Splits_File cora_hom_exact_k64 cora_hom_inexact_k64 __temporary_CORA_files/__TEMP__Aux_Parallel_Splits_File BOTH  1 ONLY_CONSTRUCT\n",
    "cd ..\n",
    "```\n",
    "\n",
    "For Seq, we used:\n",
    "```bash\n",
    "# Build\n",
    "tm seq-build-cora-exact   seqc build -release -o $TOOLS/seq/cora-exact   $TOOLS/seq/cora/hom_exact.seq\n",
    "tm seq-build-cora-inexact seqc build -release -o $TOOLS/seq/cora-inexact $TOOLS/seq/cora/hom_inexact.seq\n",
    "\n",
    "# Use 24 threads\n",
    "export OMP_NUM_THREADS=24\n",
    "# Generate exact homology table\n",
    "tm seq-cora-exact $TOOLS/seq/cora-exact   $DATA/genome.fa   $OUT/seq-cora-k64\n",
    "# Generate inexact homology table\n",
    "tm seq-cora-inexact $TOOLS/seq/cora-inexact $DATA/genome.fa 1 $OUT/seq-cora-k64\n",
    "export OMP_NUM_THREADS=1\n",
    "```\n",
    "\n",
    "Seq reported 438k homologies that were not found by CORA; CORA, on the other hand, reported 252k homologies that were not found by Seq.\n",
    "These homologies are available for inspection [here](cora-homologies.tar.bz2).\n",
    "After inspection, we discovered that the homologies not found by Seq are not true homologies due to a bug in the original version\n",
    "(for example, a reported homology between `chr6:78300426-78300505` and `chr6:78300466-78300545` is incorrect as can be seen by comparing [the first](https://genome.ucsc.edu/cgi-bin/hgTracks?db=hg19&lastVirtModeType=default&lastVirtModeExtraState=&virtModeType=default&virtMode=0&nonVirtPosition=&position=chr6%3A78300426%2D78300505&hgsid=1079285129_4USpZq0gDfCkCIMWcmtMyuKdrAMJ) and [the second](https://genome.ucsc.edu/cgi-bin/hgTracks?db=hg19&lastVirtModeType=default&lastVirtModeExtraState=&virtModeType=default&virtMode=0&nonVirtPosition=&position=chr6%3A78300466%2D78300545&hgsid=1079285129_4USpZq0gDfCkCIMWcmtMyuKdrAMJ) regionâ€” note the sequence differences at the end)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## BWA-MEM (fastmap)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We used [BWA 0.7.17](https://github.com/lh3/bwa/releases/download/v0.7.17/bwa-0.7.17.tar.bz2) with the [following patch](tools/bwa-0.7.17.patch) that reports the time needed to load an index. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "BWA was run as follows:\n",
    "```bash\n",
    "# Prepare the index\n",
    "tm bwa-index-chr1 $TOOLS/bwa-0.7.17/bwa index   $DATA/chr1.fa\n",
    "tm bwa-chr1       $TOOLS/bwa-0.7.17/bwa fastmap $DATA/chr1.fa $DATA/sample.fastq\n",
    "```\n",
    "\n",
    "Seq was run as follows:\n",
    "```bash\n",
    "# Build prefetch and no-prefetch versions of fastmap\n",
    "tm seq-build-fastmap-bn seqc build -release -o $TOOLS/seq/fastmap-bn $TOOLS/seq/bwa/fastmap_build.seq\n",
    "tm seq-build-fastmap-bp \"sed 's/^#\\@prefetch/\\@prefetch/' $TOOLS/seq/bwa/fastmap_build.seq | seqc build -release -o $TOOLS/seq/fastmap-bp -\"\n",
    "# Run both versions\n",
    "tm seq-fastmap-bn-chr1 $TOOLS/seq/fastmap-bn search $DATA/chr1.fa $DATA/sample.fastq $OUT/seq-fastmap-bn\n",
    "tm seq-fastmap-bp-chr1 $TOOLS/seq/fastmap-bp search $DATA/chr1.fa $DATA/sample.fastq $OUT/seq-fastmap-bp\n",
    "```\n",
    "\n",
    "\n",
    "Rust code was built with [rust-bio v0.32.0](https://github.com/rust-bio/rust-bio/tree/v0.32.0 with the following [patch](tools/rust/rust-bio.patch) that allows proper SMEM reconstruction:\n",
    "```bash\n",
    "# Build the tool\n",
    "(cd $TOOLS/rust; cargo build --release)\n",
    "# Run Rust fastmap\n",
    "tm rust-fastmap-chr1 $TOOLS/rust/target/release/biorust fastmap $DATA/chr1.fa $DATA/sample.fastq $OUT/rust-fastmap-chr1\n",
    "```\n",
    "\n",
    "SeqAn/C++ code was built with [SeqAn v3.0.2](https://github.com/seqan/seqan3/releases/download/3.0.2/seqan3-3.0.2-Source.tar.xz) with g++ 10:\n",
    "```bash\n",
    "# Build the tool\n",
    "( cd $TOOLS/seqan; \n",
    "  g++ -std=c++2a -fconcepts -O3 smem_seqan.cc -o fastmap \\\n",
    "    -I seqan/3.0.2/include -I seqan/3.0.2/include/seqan3/submodules/range-v3/include -I seqan/3.0.2/include/seqan3/submodules/sdsl-lite/include )\n",
    "# Run SeqAn fastmap\n",
    "tm seqan-fastmap-chr1 $TOOLS/seqan/fastmap $DATA/chr1.fa $DATA/sample.fastq\n",
    "```\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## mrsFAST\n",
    "\n",
    "We used [mrsFAST v3.4.1](https://github.com/sfu-compbio/mrsfast/tree/75b49304992302ed8754da139ccb464297fda930) with the following [patch](tools/mrsfast-3.4.1.patch)\n",
    "that turns off the calculation of MD tag for fair comparison, and that force-enabes SSE4 version to be built.\n",
    "\n",
    "```bash\n",
    "# Index the genomes\n",
    "tm mrsfast-index-hg19 $TOOLS/mrsfast/mrsfast --index  $DATA/genome.fa\n",
    "tm mrsfast-index-chr1 $TOOLS/mrsfast/mrsfast --index  $DATA/chr1.fa\n",
    "# Inexact version (e=2)\n",
    "tm mrsfast-chr1 $TOOLS/mrsfast/mrsfast --search $DATA/chr1.fa --seq $DATA/FIN1_1.fastq --crop 96 -e 2 -o $OUT/mrsfast-chr1 -u /dev/null --disable-sam-header --mem 10\n",
    "# Exact version\n",
    "tm mrsfast-hg19-all $TOOLS/mrsfast/mrsfast --search $DATA/genome.fa --seq $DATA/FIN1_1.fastq -e 0 -o $OUT/mrsfast-hg19-all -u /dev/null --disable-sam-header --mem 200\n",
    "```\n",
    "\n",
    "Seq version was run as follows:\n",
    "\n",
    "```bash\n",
    "# Build prefetch and no-prefetch versions of exact match mapper (FM-index)\n",
    "tm seq-build-mrsfast-exact-n seqc build -release -o $TOOLS/seq/mrsfast-exact-n $TOOLS/seq/mrsfast/exact.seq\n",
    "tm seq-build-mrsfast-exact-p \"sed 's/^#@prefetch/@prefetch/' $TOOLS/seq/mrsfast/exact.seq | seqc build -release -o $TOOLS/seq/mrsfast-exact-p -\"\n",
    "# Build inexact mapper (k-mer index)\n",
    "tm seq-build-mrsfast-inexact seqc build -release -o $TOOLS/seq/mrsfast-inexact $TOOLS/seq/mrsfast/mrsfast.seq\n",
    "# Index the genomes\n",
    "tm seq-mrsfast-index-hg19 $TOOLS/seq/mrsfast-exact-n index  $DATA/genome.fa\n",
    "tm seq-mrsfast-index-chr1 $TOOLS/seq/mrsfast-inexact index  $DATA/chr1.fa\n",
    "# Inexact version\n",
    "tm seq-mrsfast-chr1 $TOOLS/seq/mrsfast-inexact search $DATA/chr1.fa $DATA/FIN1_1.fastq $OUT/seq-mrsfast-chr1\n",
    "# Exact version\n",
    "tm seq-mrsfast-hg19-n $TOOLS/seq/mrsfast-exact-n search $DATA/genome.fa $DATA/FIN1_1.fastq $OUT/seq-mrsfast-hg19-n\n",
    "tm seq-mrsfast-hg19-p $TOOLS/seq/mrsfast-exact-p search $DATA/genome.fa $DATA/FIN1_1.fastq $OUT/seq-mrsfast-hg19-p\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## minimap2 (Smith-Waterman alignment)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The original [minimap2](https://github.com/lh3/minimap2) uses [KSW2](https://github.com/lh3/ksw2) library that Seq uses internally for `seq.align` function.\n",
    "Thus, we just run unoptimized `seq.align` to measure the performance of minimap2 / KSW2 alignment.\n",
    "\n",
    "```bash\n",
    "# Build the tool\n",
    "tm seq-build-sw seqc build -release -o $TOOLS/seq/sw $TOOLS/seq/minimap2/sw.seq\n",
    "# Run the benchmark. It reports the average time for: \n",
    "#  - non-optimized alignment, \n",
    "#  - SSE4-optimized inter-alignment, \n",
    "#  - AVX2-optimized inter-alignment, and \n",
    "#  - AVX512F-optimized inter-alignment (if available).\n",
    "tm seq-sw $TOOLS/seq/sw $DATA/queries_m $DATA/targets_m\n",
    "```\n",
    "\n",
    "We compared our alignment with [rust-bio v0.32.0](https://github.com/rust-bio/rust-bio/tree/v0.32.0), [SeqAn v3.0.2](https://github.com/seqan/seqan3/releases/download/3.0.2/seqan3-3.0.2-Source.tar.xz) and [Bio.jl and BioAlignments.jl v1.0.1](https://biojulia.net/Bio.jl/stable) in Julia v1.6.0.\n",
    "Other tools (SeqAn, BioJulia and rust-bio) do not support consistently advanced alignment options (e.g. setting a band width, Z-drop score and so on).\n",
    "Thus, we compared a simple version of Seq's `seq.align` to these tools to maintain the consistency:\n",
    "\n",
    "```bash\n",
    "# Build and run a simple version of the Seq alignment benchmark\n",
    "tm seq-build-sw-simple seqc build -release -o $TOOLS/seq/sw-simple $TOOLS/seq/minimap2/sw_simple.seq\n",
    "tm seq-sw-simple $TOOLS/seq/sw-simple $DATA/queries.small $DATA/targets.small\n",
    "# Build and run rust-bio alignment benchmark\n",
    "(cd $TOOLS/rust; cargo build --release)\n",
    "tm rust-sw-simple $TOOLS/rust/target/release/biorust align $DATA/queries.small $DATA/targets.small\n",
    "# Build and run SeqAn v3 alignment benchmark\n",
    "( cd $TOOLS/seqan; \n",
    "  g++ -std=c++2a -fconcepts -O3 align_seqan.cc -o sw \\\n",
    "    -I seqan/3.0.2/include -I seqan/3.0.2/include/seqan3/submodules/range-v3/include -I seqan/3.0.2/include/seqan3/submodules/sdsl-lite/include )\n",
    "tm seqan-sw-simple $TOOLS/seqan/sw $DATA/queries.small $DATA/targets.small\n",
    "# Run Julia/BioJulia alignment benchmark\n",
    "tm julia-sw-simple julia --check-bounds=no -O3 $TOOLS/julia/align.jl $DATA/queries.small $DATA/targets.small\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## AVID (global alignment)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We used a Linux version of [AVID v2.1 build 0](http://genome.lbl.gov/vista/mvista/download.shtml). (Note: you need to register to download a statically compiled binary. Source code is not available).\n",
    "\n",
    "As AVID only supports reading a single pair of sequences, we used a wrapper script [runavid.py](tools/runavid.py) to run AVID and its Seq counterpart on many pairs of sequences and collect the correct time.\n",
    "\n",
    "```bash\n",
    "# Run AVID on WGAC and SEDEF-generated segmental duplications\n",
    "tm avid-wgac  python3 $TOOLS/runavid.py $DATA/wgac-segdups  $OUT/avid-wgac  avid\n",
    "tm avid-sedef python3 $TOOLS/runavid.py $DATA/sedef-segdups $OUT/avid-sedef avid\n",
    "```\n",
    "\n",
    "Seq was run as follows:\n",
    "```bash\n",
    "# Build Seq-AVID\n",
    "tm seq-build-sw seqc build -release -o $TOOLS/seq/sw $TOOLS/seq/avid/avid.seq\n",
    "# Run Seq-AVID on WGAC and SEDEF-generated segmental duplications\n",
    "tm seq-avid-wgac  python3 $TOOLS/runavid.py $DATA/wgac-segdups  $OUT/seq-avid-wgac  seq\n",
    "tm seq-avid-sedef python3 $TOOLS/runavid.py $DATA/sedef-segdups $OUT/seq-avid-sedef seq\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## UMI-tools (barcode whitelisting)\n",
    "\n",
    "We used [UMI-tools v1.1.1](https://github.com/CGATOxford/UMI-tools/tree/1.1.1):\n",
    "\n",
    "```bash\n",
    "tm umi-tools umi_tools whitelist --stdin $DATA/hgmm_100_R1.fastq --bc-pattern=CCCCCCCCCCCCCCCCNNNNNNNNNN --log2stderr\n",
    "```\n",
    "\n",
    "Seq verson was run as follows:\n",
    "```bash\n",
    "# Build\n",
    "tm seq-build-umi seqc build -release -o $TOOLS/seq/umi $TOOLS/seq/umi-tools/whitelist.seq\n",
    "# Run\n",
    "tm seq-umi $TOOLS/seq/umi $DATA/hgmm_100_R1.fastq\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## GATK (SplitNCigar)\n",
    "\n",
    "We used [GATK 4.1.4.1](https://github.com/broadinstitute/gatk/releases/download/4.1.4.1/gatk-4.1.4.1.zip) to run `SplitNCigar` as follows:\n",
    "```bash\n",
    "for i in giab k562 cytosol; do\n",
    "  tm gatk-$i $TOOLS/gatk-4.1.4.1/gatk SplitNCigarReads -I $DATA/$i.gatk.bam -O $OUT/gatk-$i.sam -R $DATA/Homo_sapiens_assembly19_1000genomes_decoy.fasta\n",
    "done\n",
    "```\n",
    "\n",
    "Seq was run as follows:\n",
    "```bash\n",
    "tm seq-build-gatk seqc build -release -o $TOOLS/seq/gatk $TOOLS/seq/gatk-splitncigar/gatk-split.seq\n",
    "for i in giab k562 cytosol; do\n",
    "  tm seq-gatk-$i $TOOLS/seq/gatk $DATA/$i.gatk.bam $DATA/Homo_sapiens_assembly19_1000genomes_decoy.fasta $OUT/gatk-$i.sam\n",
    "done\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "",
   "name": ""
  },
  "language_info": {
   "name": ""
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
